pkg/operator/ceph/pool/controller.go

```diff
@@ -18,155 +18,219 @@ limitations under the License.
-	if err := createPool(c.context, pool); err != nil {
-		logger.Errorf("failed to create (modify) pool %q. %v", pool.ObjectMeta.Name, err)
-		updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.FailedStatus, c.context)
-		return
+	// CREATE or UPDATE
+	err = r.reconcile(cephBlockPool, cephClusterSpec)
+	if err != nil {
+		logger.Errorf("failed to reconcile. %v", err)
+		updateCephBlockPoolStatus(cephBlockPool.GetName(), cephBlockPool.GetNamespace(), k8sutil.ReconcileFailedStatus, r.context)
+		return immediateRetryResult, nil
 	}
-	updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.ReadyStatus, c.context)
-}
 
-// ParentClusterChanged determines wether or not a CR update has been sent
-func (c *PoolController) ParentClusterChanged(cluster cephv1.ClusterSpec, clusterInfo *cephconfig.ClusterInfo, isUpgrade bool) {
-	logger.Debugf("No need to update the pool after the parent cluster changed")
+	return reconcile.Result{}, nil
 }
 
-func poolChanged(old, new cephv1.PoolSpec) bool {
-	if old.Replicated.Size != new.Replicated.Size {
-		logger.Infof("pool replication changed from %d to %d", old.Replicated.Size, new.Replicated.Size)
-		return true
-	}
-	return false
-}
+func (r *ReconcileCephBlockPool) reconcile(pool *cephv1.CephBlockPool, cephClusterSpec cephv1.ClusterSpec) error {
+	// Create pools
+	if pool.Status.Phase != k8sutil.ReadyStatus {
+		updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.ProcessingStatus, r.context)
+		err := createPool(r.context, pool)
+		if err != nil {
+			updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.FailedStatus, r.context)
+			return errors.Wrapf(err, "failed to create pool %q.", pool.GetName())
+		}
 
-func (c *PoolController) onDelete(obj interface{}) {
-	if c.clusterSpec.External.Enable && c.clusterSpec.CephVersion.Image == "" {
-		logger.Warningf("Deleting pools for an external ceph cluster is disabled because no Ceph image is specified")
-		return
-	}
+		// Set Ready status
+		updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.ReadyStatus, r.context)
 
-	pool, err := getPoolObject(obj)
-	if err != nil {
-		logger.Errorf("failed to get pool object. %v", err)
-		return
+		// Let's return here so that on the initial creation we don't check for update right away
+		return nil
 	}
-	if err := deletePool(c.context, pool); err != nil {
-		logger.Errorf("failed to delete pool %q. %v", pool.ObjectMeta.Name, err)
+
+	// Verify whether the pool needs to be updated
+	if pool.Status.Phase == k8sutil.ReadyStatus {
+		needsUpdate, err := r.needsUpdate(pool)
+		if err != nil {
+			return errors.Wrapf(err, "failed to check if the Ceph Block Pool CR %q needs to be updated", pool.GetName())
+		}
+		if needsUpdate {
+			if cephClusterSpec.External.Enable && cephClusterSpec.CephVersion.Image == "" {
+				logger.Warningf("updating pools for an external ceph cluster is disabled because no ceph image is specified")
```

Comment:
Why do pools need a ceph image to be updated?

Reply:
With external cluster, I believe we do some version check for certain things/

Reply:
I see the code just moved around, but it doesn't seem like a necessary check for pools since we don't start any daemons for pools.

Reply:
removing this simplifies a bit.